syscall/devices/hostepoll/hostepoll.c

```diff
@@ -457,26 +465,32 @@ static int _epoll_wait(
         if (retval > maxevents)
             OE_RAISE_ERRNO(OE_EINVAL);
 
+        locked = true;
+        oe_mutex_lock(&epoll->lock);
+
         for (int i = 0; i < retval; i++)
         {
-            struct oe_epoll_event* event = &events[i];
-            const mapping_t* mapping;
+            struct oe_epoll_event* const event = &events[i];
+            const mapping_t* const mapping = _map_find(epoll, event->data.fd);
 
-            oe_spin_lock(&epoll->lock);
+            if (mapping)
+                event->data.u64 = mapping->event.data.u64;
+            else
             {
-                if ((mapping = _map_find(epoll, event->data.fd)))
-                    event->data.u64 = mapping->event.data.u64;
+                // fd has been deleted between the return of epoll_wait and the
+                // acquisition of the lock.
+                --retval;
```

Comment:
Do we also need to decrement `i` here? I think if we don't we'll end up skipping the next event.

Reply:
Good catch! Fixed.